ProjFS.Mac/PrjFSLib/PrjFSUser.cpp

```diff
@@ -32,50 +70,77 @@ io_connect_t PrjFSService_ConnectToDriver(enum PrjFSServiceUserClientType client
+        return IO_OBJECT_NULL;
     }
-    
-    kextVersionString = static_cast<CFStringRef>(kextVersionObj);
-    if (kCFCompareEqualTo != CFStringCompare(kextVersionString, CFSTR(PrjFSKextVersion), 0 /* options: case sensitive */))
+
+    kern_return_t result = IOServiceOpen(prjfsService, mach_task_self(), clientType, &connection);
+    IOObjectRelease(prjfsService);
+    if (kIOReturnSuccess != result || IO_OBJECT_NULL == connection)
     {
-        const char* kextVersion = CFStringGetCStringPtr(kextVersionString, kCFStringEncodingUTF8) ?: "???";
-        std::cerr << "PrjFS kernel service interface version mismatch. Kernel: " << kextVersion << ", this library expects: " << PrjFSKextVersion << std::endl;
-        goto CleanupAndFail;
+        std::cerr << "Failed to open connection to kernel service; error: 0x" << std::hex << result << ", connection 0x" << std::hex << connection << std::endl;
+        connection = IO_OBJECT_NULL;
     }
-    else
+
+    return connection;
+}
+
+static void ServiceMatched(
+    void* refcon,
+    io_iterator_t iterator)
+{
+    PrjFSService_WatchContext* context = static_cast<PrjFSService_WatchContext*>(refcon);
+    
+    while (io_service_t prjfsService = IOIteratorNext(context->notificationIterator))
     {
-        CFRelease(kextVersionString);
-        
-        // Version matches, connect to kernel service
-        kern_return_t result = IOServiceOpen(prjfsService, mach_task_self(), clientType, &connection);
-        IOObjectRelease(prjfsService);
-        if (kIOReturnSuccess != result || IO_OBJECT_NULL == connection)
+        io_connect_t connection = IO_OBJECT_NULL;
+        if (PrjFSService_ValidateVersion(prjfsService))
         {
-            std::cerr << "Failed to open connection to kernel service: 0x" << std::hex << result << ", connection 0x" << std::hex << connection << std::endl;
-            connection = IO_OBJECT_NULL;
+            kern_return_t result = IOServiceOpen(prjfsService, mach_task_self(), context->clientType, &connection);
+            if (result != kIOReturnSuccess)
+            {
+                std::cerr << "Failed to open connection to kernel service; error: 0x" << std::hex << result << ", connection 0x" << std::hex << connection << std::endl;
+            }
         }
+        
+        context->discoveryCallback(prjfsService, connection, context);
+        
+        IOObjectRelease(prjfsService);
     }
-    
-    return connection;
+}
 
-CleanupAndFail:
-    if (IO_OBJECT_NULL != prjfsService)
+PrjFSService_WatchContext* PrjFSService_WatchForServiceAndConnect(
+    IONotificationPortRef notificationPort,
+    enum PrjFSServiceUserClientType clientType,
+    std::function<void(io_service_t, io_connect_t, PrjFSService_WatchContext*)> discoveryCallback)
+{
+    CFDictionaryRef matchDict = IOServiceMatching(PrjFSServiceClass);
+    PrjFSService_WatchContext* context = new PrjFSService_WatchContext { std::move(discoveryCallback), IO_OBJECT_NULL, clientType };
```

Comment:
I contemplated that, but it makes the API either less explicit or more complex; we'd have to return a `shared_ptr<PrjFSService_WatchContext>` and move the logic for `PrjFSService_StopWatching()` into the destructor for `PrjFSService_WatchContext` so that we stop watching when the last reference disappears, or keep track of `PrjFSService_WatchContext` objects past the point of `PrjFSService_StopWatching()` being called. With the current approach, we can treat `PrjFSService_WatchContext` as completely opaque to API users, which suits our C-with-some-C++-features approach nicely in my opinion.